### DISCUSSION TITLE
gnark vs zokrates

### DISCUSSION THREAD

**Discussion by adust09** [Consensys/gnark#119] [Q&A] [answered]
Hi, I am designing a module for Cosmos Blockchain using ganrk.
I am comparing various zk-SNARKs libraries, and I am noticing that gnark works well with Cosmos since it is implemented in Go.
In this context, I have a few questions. (Mainly, how it compares to zokrates.)

Q1. Is there any difference from zokrates, which is an Ethereum-specific library?

Q2. I'm aware that zokrates has a specification to automatically drop smart contract functions into the circuit (or not?).
Is there such a specification in gnark? Or, is there any such research going on in the future?

Q3. In gnark, you need to describe the constraints of the circuit by yourself, right?
From the tutorials, it seems that this is limited to constraints on finite bodies, but what kind of circuits are practical to apply to existence proofs such as those used in NFT and L2?

**Comment by gbotrel** [comment, marked as answer]
Hello :-) 

Here are few differences between `gnark` and Zokrates; 
1. `gnark` has no external dependency. It is not a wrapper around a backend, it implements the backend and the frontend part of a zkSNARK scheme. Performances are good ([1](https://docs.gnark.consensys.net/en/latest/#gnark-is-fast) & [2](https://hackmd.io/@zkteam/eccbench))
2. `gnark` circuits are [written in Go](https://docs.gnark.consensys.net/en/latest/#gnark-circuits-are-written-in-go). It is a design choice not to propose another zkSNARK-DSL, and rely on a robust toolchain (Go). Among other things, it enables to debug, document, test circuits like any other Go programs. Easy IDE integration / no long term maintenance effort. Easier to integrate into larger project (APIs, serialization, logging, monitoring, ...). 
3. `gnark` is Apache 2.0 (vs LGPL)
4. `gnark` supports Groth16 and PlonK proving schemes

on Q2;
Integration with Ethereum is minimalist at the moment; you can [export a VerifyingKey into a solidity smart contract](https://docs.gnark.consensys.net/en/latest/HowTo/prove/#verify-a-proof-on-ethereum) . [Here is a code sample](https://github.com/gbotrel/gnark-workshop). 

on Q3: 
Yes, currently `gnark` frontend API is "low level" and works with the same finite field object as the backend curve. 

Our main effort in the upcoming months is to enable programmable rollups.

**Reply by adust09** [reply]
Sorry for the late reply. Thank you very much for your very kind answer.

In addition to your answer, actually touching gnark and the other zk library (snarkjs) has helped me understand gnark better.

Especially the fact that it's implemented in Go, which seems like a good fit for a Go-based blockchain like Cosmos!

**Reply by yzhanginwa** [reply]
How is your cosmos module project going? I'm interested in doing something similar. If you've done and made it open, I would like to give it a try. Thanks!

**Reply by adust09** [reply]
Thank you for your comment, but I'm pending this project now. Sorry.
